src/common/sockman.h

```diff
@@ -0,0 +1,427 @@
+
+            const fs::path private_key_file;
+            const Proxy sam_proxy;
+        };
+
+        /**
+         * I2P options. If set then a thread will be started that will accept incoming I2P connections.
+         */
+        std::optional<I2P> i2p;
+    };
+
+    /**
+     * Start the necessary threads for sockets IO.
+     */
+    void StartSocketsThreads(const Options& options);
+
+    /**
+     * Join (wait for) the threads started by `StartSocketsThreads()` to exit.
+     */
+    void JoinSocketsThreads();
+
+    /**
+     * A more readable std::tuple<std::string, uint16_t> for host and port.
+     */
+    struct StringHostIntPort {
+        const std::string& host;
+        uint16_t port;
+    };
+
+    /**
+     * Make an outbound connection, save the socket internally and return a newly generated node id.
+     * @param[in] to The address to connect to, either as CService or a host as string and port as
+     * an integer, if the later is used, then `proxy` must be valid.
+     * @param[in] is_important If true, then log failures with higher severity.
+     * @param[in] proxy Proxy to connect through if `proxy.IsValid()` is true.
+     * @param[out] proxy_failed If `proxy` is valid and the connection failed because of the
+     * proxy, then it will be set to true.
+     * @param[out] me If the connection was successful then this is set to the address on the
+     * local side of the socket.
+     * @return Newly generated node id, or std::nullopt if the operation fails.
+     */
+    std::optional<NodeId> ConnectAndMakeNodeId(const std::variant<CService, StringHostIntPort>& to,
+                                               bool is_important,
+                                               const Proxy& proxy,
+                                               bool& proxy_failed,
+                                               CService& me)
+        EXCLUSIVE_LOCKS_REQUIRED(!m_connected_mutex, !m_unused_i2p_sessions_mutex);
+
+    /**
+     * Disconnect a given peer by closing its socket and release resources occupied by it.
+     * @return Whether the peer existed and its socket was closed by this call.
+     */
+    bool CloseConnection(NodeId node_id)
+        EXCLUSIVE_LOCKS_REQUIRED(!m_connected_mutex);
+
+    /**
+     * Try to send some data to the given node.
+     * @param[in] node_id Identifier of the node to send to.
+     * @param[in] data The data to send, it might happen that only a prefix of this is sent.
+     * @param[in] will_send_more Used as an optimization if the caller knows that they will
+     * be sending more data soon after this call.
+     * @param[out] errmsg If <0 is returned then this will contain a human readable message
+     * explaining the error.
+     * @retval >=0 The number of bytes actually sent.
+     * @retval <0 A permanent error has occurred.
+     */
+    ssize_t SendBytes(NodeId node_id,
+                      Span<const unsigned char> data,
```

Comment:
Why not use std::span? I thought `Span` was only a temporary fill-in until C++20 something something? Also, I'm using std::span elsewhere... is `Span` the preference?

Reply:
Even in relatively new code folks often use `Span`, e.g. #30043. cc @sipa, @laanwj

Reply:
Must have been just inertia. The caller of the newly added `SendBytes()` has an argument that is already `Span` from elsewhere in the code, coming from here:

https://github.com/bitcoin/bitcoin/blob/c1252b14d714448295200a595086dd3e78b46c8f/src/net.h#L307-L311

Now I changed the argument of the newly added `SendBytes()` to `std::span`, looks better and the `Span` variable is accepted. Thanks!

wen `s/Span/std::span/` all over the place?

Reply:
> wen `s/Span/std::span/` all over the place?

now: https://github.com/bitcoin/bitcoin/pull/31519